src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1037 @@
+    for (const auto& fd_node : fd_node_list_) {
+      if (!fd_node->already_shutdown) {
+        GRPC_ARES_WRAPPER_TRACE_LOG(
+            "request:%p OnAresBackupPollAlarm; ares_process_fd. fd=%s", this,
+            fd_node->polled_fd->GetName());
+        ares_socket_t as = fd_node->polled_fd->GetWrappedAresSocketLocked();
+        ares_process_fd(channel_, as, as);
+      }
+    }
+    if (!shutting_down_) {
+      EventEngine::Duration next_ares_backup_poll_alarm_duration =
+          calculate_next_ares_backup_poll_alarm_duration();
+      ares_backup_poll_alarm_handle_ = event_engine_->RunAfter(
+          next_ares_backup_poll_alarm_duration,
+          [self = Ref(DEBUG_LOCATION, "OnAresBackupPollAlarm")]() {
+            self->OnAresBackupPollAlarm();
+          });
+    }
+    Work();
+    if (shutting_down_) {
+      OnResolveLockedAndUnlock();
+      return;
+    }
+  }
+  mu().Unlock();
+}
+
+void GrpcAresRequest::ShutdownPolledFdsLocked(absl::Status status) {
+  for (const auto& fd_node : fd_node_list_) {
+    if (!fd_node->already_shutdown) {
+      GRPC_ARES_WRAPPER_TRACE_LOG("request: %p shutdown fd: %s", this,
+                                  fd_node->polled_fd->GetName());
+      fd_node->polled_fd->ShutdownLocked(status);
+      fd_node->already_shutdown = true;
+    }
+  }
+}
+
+absl::StatusOr<GrpcAresHostnameRequest*> GrpcAresHostnameRequest::Create(
+    absl::string_view name, absl::string_view default_port,
+    absl::string_view dns_server, EventEngine::Duration timeout,
+    std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory,
+    EventEngine* event_engine) ABSL_NO_THREAD_SAFETY_ANALYSIS {
+  auto* request = new GrpcAresHostnameRequest(
+      name, default_port, timeout, std::move(polled_fd_factory), event_engine);
+  absl::StatusOr<std::string> result = request->ParseNameToResolve();
+  if (!result.ok()) {
+    delete request;
+    return result.status();
+  }
+  absl::Status status = request->ParsePort(*result);
+  if (!status.ok()) {
+    delete request;
+    return status;
+  }
+  status = request->InitializeAresOptions(dns_server);
+  if (!status.ok()) {
+    delete request;
+    return status;
+  }
+  return request;
+}
+
+void GrpcAresHostnameRequest::Start(
+    absl::AnyInvocable<void(absl::StatusOr<Result>)> on_resolve) {
+  grpc_core::ReleasableMutexLock lock(&mu());
+  GPR_ASSERT(initialized());
+  on_resolve_ = std::move(on_resolve);
+  GRPC_ARES_WRAPPER_TRACE_LOG(
+      "request:%p c-ares GrpcAresHostnameRequest::Start name=%s, "
+      "default_port=%s",
+      this, name().c_str(), default_port_.c_str());
+  // Early out if the target is an ipv4 or ipv6 literal.
+  if (ResolveAsIPLiteralLocked()) {
+    auto finish = [on_resolve = std::move(on_resolve_),
+                   result = std::move(result_),
+                   self = Ref(DEBUG_LOCATION, "Start")]() mutable {
+      on_resolve(std::move(result));
+    };
+    lock.Release();
+    event_engine_->Run(std::move(finish));
```

Comment:
That is not allowed today per the API spec. It could lead to hairy situations with held locks and whatnot if that becomes non-deterministic.

Reply:
In the new implementation, for now we are relying on `event_engine_->Run` to run the callback on another thread which does not hold the lock on the AresResolver's mutex (not running the callback inline on the same thread) so that the lock order will not be reversed.

For example:
```
void AresResolver::OnHostbynameDoneLocked(void* arg, int status,
                                          int /*timeouts*/,
                                          struct hostent* hostent) {
  std::unique_ptr<HostnameQueryArg> hostname_qa(
      static_cast<HostnameQueryArg*>(arg));
  auto* ares_resolver = hostname_qa->ares_resolver;
  auto nh = ares_resolver->callback_map_.extract(hostname_qa->callback_map_id);
  GPR_ASSERT(!nh.empty());
  GPR_ASSERT(
      absl::holds_alternative<EventEngine::DNSResolver::LookupHostnameCallback>(
          nh.mapped()));
  auto callback = absl::get<EventEngine::DNSResolver::LookupHostnameCallback>(
      std::move(nh.mapped()));
  if (status != ARES_SUCCESS) {
    std::string error_msg =
        absl::StrFormat("c-ares status is not ARES_SUCCESS name=%s: %s",
                        hostname_qa->qname.c_str(), ares_strerror(status));
    GRPC_ARES_RESOLVER_TRACE_LOG("request:%p on_hostbyname_done_locked: %s",
                                 ares_resolver, error_msg.c_str());
    ares_resolver->event_engine_->Run(
        [callback = std::move(callback),
         status = AresStatusToAbslStatus(status, error_msg)]() mutable {
          callback(status);
        });
    return;
  }
......
```